File: src/solverFixedProd.jl
using JuMP, Gurobi, HiGHS
include("default_values.jl")


"""
## Arguments
- wind_capa : Capacity of the wind farm (MW)
- wind_profile : Wind profile (% of the capacity)
- solar_capa : Capacity of the solar farm (MW)
- solar_profile : Solar profile (% of the capacity)
- demand : Demand of hydrogen (Kg)
- price_grid : Price of the grid (€ / MWh)
- price_curtailing : Price of the curtailed energy (€ / MWh)
- prodChangePenality : Penality for changing the production (€ / times the production changes)
- initCharge : Initial charge of the battery (MWh)
- initStock : Initial stock of the tank (Kg)
- finalStock : Final stock of the tank (Kg), if None, the final stock is not constrained
- finalCharge : Final charge of the battery (MWh), if None, the final charge is not constrained
## Returns a dictionnary with following keys
- battery_capa : Capacity of the battery (MWh)
- tank_capa : Capacity of the tank (Kg)
- charge : Battery charge (MWh)
- prod : Production of hydrogen (Kg)
- stock : Stock of hydrogen in the tank (Kg)
- elecGrid : Electricity consumption from the grid (MWh)
- consPPA : Consumption of PPA energy (MWh)
- flowBat : Flow of energy to / from the battery ( > 0 if charging, < 0 if discharging) (MWh)
- flowH2 : Flow of hydrogen to / from the tank ( > 0 if charging, < 0 if discharging) (Kg)
- operating_cost : Operating cost (€)
- storage_cost : Construction cost (€) # In this version, could be computed before optimization
"""
function solveFixedProd(
    windCapa :: Float64,
    solarCapa :: Float64,
    windProfile :: Array{Float64, 1},
    solarProfile :: Array{Float64, 1},
    demand :: Float64,
    price_grid = PRICE_GRID,
    price_curtailing = PRICE_CURTAILING,
    price_penality = 0.,
    initCharge =  0.,
    initStock = 0.,
    finalCharge = missing,
    finalStock = missing,
)
    # Number of time steps
    T = length(windProfile)
    if length(solarProfile) != T
        throw(ArgumentError("The length of the solar profile should be equal to the length of the wind profile"))
    end

    # Create the model
    model = Model(HiGHS.Optimizer)
    # Storage variables
    batteryCapa = @variable(model, lower_bound = 0.)
    tankCapa = @variable(model, lower_bound = 0.)
    # Main variables
    charge = @variable(model, [1:T+1], lower_bound = 0.)
    stock = @variable(model, [1:T+1], lower_bound = 0.)
    prod = @variable(model, [1:T], lower_bound = 0.)
    # Electricity consumption
    elecPPA = @variable(model, [1:T], lower_bound = 0.)
    elecGrid = @variable(model, [1:T], lower_bound = 0.)
    curtailing = @variable(model, [1:T], lower_bound = 0.)
    # Flow of elecricity / hydrogen
    flowBat = @variable(model, [1:T])
    flowH2 = @variable(model, [1:T])
    # Production changing penality
    prodHasChanged = @variable(model, [2:T], Bin)
    # Costs pseudo-variables
    operating_cost = @variable(model)
    storage_cost = @variable(model)

    # Initial charge & stock
    @constraint(model, charge[1] == initCharge)
    @constraint(model, stock[1] == initStock)
    # Final charge & stock
    if !ismissing(finalCharge)
        @constraint(model, charge[T+1] == finalCharge)
    end
    if !ismissing(finalStock)
        @constraint(model, stock[T+1] == finalStock)
    end
    # Get the per hour discharge of the batteryn from the per month parameter
    perHourDischarge = EBAT ^ (1 / (30 * 24))
    # PPA contract
    @constraint(model, [t ∈ 1:T], elecPPA[t] == windProfile[t] * windCapa + solarProfile[t] * solarCapa)
    # Demand satisfaction
    @constraint(model, [t ∈ 1:T], prod[t] == flowH2[t] + demand)
    # Electricity consumption
    @constraint(model, [t ∈ 1:T], elecGrid[t] + elecPPA[t] - curtailing[t] == prod[t] * EELEC + flowBat[t])
    # Battery charge
    @constraint(model, [t ∈ 1:T], charge[t+1] == perHourDischarge * charge[t] + flowBat[t])
    # Tank stock
    @constraint(model, [t ∈ 1:T], stock[t+1] == 1 * stock[t] + flowH2[t])
    # Flow of electricity / hydrogen
    @constraint(model, [t ∈ 1:T], -FBAT <= flowBat[t] <= FBAT)
    # Electrolyzer consumption
    @constraint(model, [t ∈ 1:T], prod[t] * EELEC <= CELEC)
    # Maximum charge & stock
    @constraint(model, [t ∈ 1:T+1], charge[t] <= batteryCapa)
    @constraint(model, [t ∈ 1:T+1], stock[t] <= tankCapa)
    # Boolean variable for production change
    @constraint(model, [t ∈ 2:T], prod[t] - prod[t-1] <= 2 * CELEC * prodHasChanged[t] / EELEC)
    @constraint(model, [t ∈ 2:T], prod[t-1] - prod[t] <= 2 * CELEC * prodHasChanged[t] / EELEC)
    # Operating cost
    @constraint(model, operating_cost == 
    sum(elecGrid) * price_grid
     + sum(curtailing) * price_curtailing 
     + sum(prodHasChanged) * price_penality
    )
    # Storage cost
    @constraint(model, storage_cost == COST_BAT * batteryCapa + COST_TANK * tankCapa)
    # Objective
    @objective(model, Min, operating_cost + storage_cost)
    optimize!(model)

    return Dict(
        "battery_capa" => value.(batteryCapa),
        "tank_capa" => value.(tankCapa),
        "charge" => value.(charge),
        "prod" => value.(prod),
        "stock" => value.(stock),
        "elecGrid" => value.(elecGrid),
        "curtail" => value.(curtailing),
        "elecPPA" => value.(elecPPA),
        "flowBat" => value.(flowBat),
        "flowH2" => value.(flowH2),
        "operating_cost" => value(operating_cost),
        "storage_cost" => value(storage_cost),
    )
end

File: old_python/src/solverFixedProd.py
from pulp import *
from src.default_values import *


def solve_fixed_prod(
    wind_capa,
    wind_profile,
    solar_capa,
    solar_profile,
    demand,
    ebat=EBAT,
    fbat=FBAT,
    eelec=EELEC,
    celec=CELEC,
    initCharge=0,
    initStock=0,
    finalStock=None,
    finalCharge=None,
    price_grid = PRICE_GRID,
    cost_bat = COST_BAT,
    cost_tank = COST_TANK
):
    """
    Parameters ::
    - wind_capa : Capacity of the wind farm (MW)
    - wind_profile : Wind profile (% of the capacity)
    - solar_capa : Capacity of the solar farm (MW)
    - solar_profile : Solar profile (% of the capacity)
    - battery_capa : Capacity of the battery (MWh)
    - tank_capa : Capacity of the tank (Kg)
    - demand : Demand of hydrogen (Kg)
    - ebat : Efficiency of the battery, in charge loss per month
    - fbat : Maximum flow of energy to / from the battery (MW)
    - eelec : Electricity consumption for 1 Kg of hydrogen (MWh / Kg)
    - celec : Maximum electricity consumption of the electrolyzer (MW)
    - initCharge : Initial charge of the battery (MWh)
    - initStock : Initial stock of the tank (Kg)
    - finalStock : Final stock of the tank (Kg), if None, the final stock is not constrained
    - finalCharge : Final charge of the battery (MWh), if None, the final charge is not constrained
    - price_grid : Price of the grid (€ / MWh), high to avoid using the grid / curtailing
    - cost_bat : Cost of the battery (€ / MWh)
    - cost_tank : Cost of the tank (€ / Kg) \n
    Returns a dictionnary with following keys ::
    - charge : Battery charge (MWh)
    - prod : Production of hydrogen (Kg)
    - stock : Stock of hydrogen in the tank (Kg)
    - elecGrid : Electricity consumption from the grid (MWh)
    - consPPA : Consumption of PPA energy (MWh)
    - flowBat : Flow of energy to / from the battery ( > 0 if charging, < 0 if discharging) (MWh)
    - flowH2 : Flow of hydrogen to / from the tank ( > 0 if charging, < 0 if discharging) (Kg)
    - operating_cost : Operating cost (€)
    - storage_cost : Construction cost (€) # In this version, could be computed before optimization
    """
    # Create the model
    problem = LpProblem("Production_Problem", LpMinimize)
    T = len(wind_profile)
    if T != len(solar_profile):
        raise ValueError("The wind and solar profiles must have the same length")
    indices = range(T)
    indicesExt = range(T + 1)  # Add the first hour of the period (Final state)
    # Variables
    charge = LpVariable.dicts("charge", indicesExt, 0, None)  # Battery charge (MWh)
    prod = LpVariable.dicts("prod", indices, 0, None)  # Production of hydrogen (Kg)
    stock = LpVariable.dicts(
        "stock", indicesExt, 0, None
    )  # Stock of hydrogen in the tank (Kg)
    elecGrid = LpVariable.dicts(
        "elec", indices, 0, None
    )  # Electricity consumption  from the grid (MWh)
    curtail = LpVariable.dicts(
        "curtail", indices, 0, None
    )  # Curtailing (MWh)
    consPPA = LpVariable.dicts(
        "consPPA", indices, 0, None
    )  # Consumption of PPA energy (MWh)
    flowBat = LpVariable.dicts(
        "flowBat", indices, None, None
    )  # Flow of energy to / from the battery ( > 0 if charging, < 0 if discharging) (MWh)
    flowH2 = LpVariable.dicts(
        "flowH2", indices, None, None
    )  # Flow of hydrogen to / from the tank ( > 0 if charging, < 0 if discharging) (Kg)
    operating_cost = LpVariable("operating_cost", 0, None)  # Operating cost (€)
    # Capacities for the storage
    battery_capa = LpVariable("battery_capa", 0, None)  # Battery capacity (MWh)
    tank_capa = LpVariable("tank_capa", 0, None)  # Tank capacity (Kg)
    storage_cost = LpVariable("storage_cost", 0, None)  # Construction cost (€)
    # Convert the per month discharge loss to per hour
    ebatPerHour = ebat ** (1 / (30 * 24))
    # Initial state
    problem += charge[0] == initCharge
    problem += stock[0] == initStock

    for t in range(T):
        # PPA constraints, we need to respect the PPA contract
        problem += consPPA[t] == wind_profile[t] * wind_capa + solar_profile[t] * solar_capa
        # Demand satisfaction
        problem += prod[t] == demand + flowH2[t]
        # Electricity consumption
        problem += consPPA[t] + elecGrid[t] - curtail[t] == eelec * prod[t] + flowBat[t]
        # Battery constraints
        problem += ebatPerHour * charge[t] + flowBat[t] == charge[t + 1]
        # Tank constraints
        problem += stock[t] + flowH2[t] == stock[t + 1]
        # Flow constraints
        problem += flowBat[t] <= fbat
        problem += -flowBat[t] <= fbat
        # Electrolyzer constraints
        problem += prod[t] * eelec <= celec

    # We also need to respect those constraints for the final state
    for t in range(T + 1):
        # Capacity constraints for battery and tank
        problem += charge[t] <= battery_capa
        problem += stock[t] <= tank_capa

    # Final state constraints
    if finalStock is not None:
        problem += stock[T] == finalStock

    if finalCharge is not None:
        problem += charge[T] == finalCharge

    # Cost functions
    problem += storage_cost == battery_capa * cost_bat + tank_capa * cost_tank
    problem += operating_cost == sum(elecGrid[t] * price_grid + curtail[t] * price_grid for t in indices)
    # Objective function
    problem += operating_cost + storage_cost

    # Solve the problem
    problem.solve()

    # Return the results

    return {
        "charge":  [charge[t].varValue for t in indicesExt],
        "prod": [prod[t].varValue for t in indices],
        "stock":  [charge[t].varValue for t in indicesExt],
        "elecGrid":  [elecGrid[t].varValue for t in indicesExt],
        "curtail": [curtail[t].varValue for t in indices],
        "consPPA":  [consPPA[t].varValue for t in indicesExt],
        "flowBat": [flowBat[t].varValue for t in indices],
        "flowH2": [flowH2[t].varValue for t in indices],
        "operating_cost": value(operating_cost),
        "storage_cost": value(storage_cost),
    }

